src/preact/use-intersect.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {useLayoutEffect, useRef, useState} from './index';
+
+/**
+ * @param {object} ref
+ * @return {?Array<IntersectionObserverEntry>}
+ */
+export function useIntersect(ref) {
```

Comment:
This is a side-effect and thus should be defined as such. This is a nuanced distinction. But the easiest way to think about this is this: if you were rendering this server-side, would you possibly have a state for this? If not - this is definitely a side-effect. So:

1. Remove the state.
2. Supply and call the handler from the args.
3. Name it `useIntersect[ion]Effect`.

Reply:
@dvoytenko, should all hooks with side effects end in `Effect`? E.g. should `useResourcesNotify` be renamed to `useResourcesNotifyEffect`?

Reply:
That would be my preference. Though `useResourceNotify` is its own case which we will be dealing with soon. But for consistency, I'd prefer that the effects were named with "Effect" suffix.

Reply:
Hmm. It feels to me like the `Timeago` component should be able to use a hook called `useRenderOnEnterViewport`. That would let Timeago have almost no logic:
```js
function Timeago(props) { 
  const ref = useRef(null); 
  useRenderOnEnterViewport(ref.current); 
  useResourcesNotify(); 

  return createElement("time", { datetime: props["datetime"], ref }, getFuzzyTimestampValue(props));
}
```

Reply:
A separate hook that makes this even simpler is ok by me, but the initial effect-hook should exist. FWIW, "render when in viewport" is very rare and also not by itself sufficient. A more common scenario could be "animate when in viewport", "play when in viewport", etc - which are not rendering features by themselves.

> useRenderOnEnterViewport(ref.current); 

This will miss the first render function. Probable should be `useRenderOnEnterViewport(ref)`?

Reply:
It seems to me that the common scenario is abstractly similar to `useIntersect(ref)`. In that case I'll restore a version of that hook and use it in `useOnEnterViewport(ref, effect)`. Originally I had `useRenderOnEnterViewport(ref)` which called the former [`useRerenderer`](https://github.com/ampproject/amphtml/pull/26507#discussion_r372625894), but this led to unclear consequences such as rerendering on both enter and exit viewport and interfering with `useResourcesNotify`.